### ISSUE TITLE
select2 not working in Jhipster application

### ISSUE THREAD

**Issue by xCozmox** [surveyjs/widgets#111]
Hi all,

i created a jhipster application in which I use surveyjs. 
I try to get the select2 dropdown working.

What I did so far:

1. npm install surveyjs-widgets
2. 
```
import * as Survey from 'survey-angular';

this.survey = new Survey.Survey();
        this.surveyCountry = this.survey.addNewPage('Country');
       this.surveyCountry.addNewQuestion('dropdown', 'country');
        const modificationCountry = this.survey.getQuestionByName('country');
        modificationCountry .type = 'dropdown';
        modificationCountry .renderAs = 'select2';
        modificationCountry .isRequired = true;
        modificationCountry .title = 'Please select:';
        modificationCountry .choices = ['Andorra', 'Germany', 'Italy'];

        Survey.Survey.cssType = 'bootstrap';
        //Survey.StylesManager.applyTheme('bootstrap');
        Survey.SurveyNG.render('surveyElement', { model: this.survey });
```
3. Nothing works. I have always a normal dropdown list.

What do I miss? Do I have to register the surveyjs-widgets somewhere in the application?

Thanks a lot for your help!

**Comment by tsv2013** [comment]
Have you seen this sample? - https://surveyjs.io/Examples/Library/?id=custom-widget-select2&platform=jQuery&theme=default

**Comment by xCozmox** [comment]
@tsv2013 sorry I updated my ticket. I accidentally submitted to early.  
And yes I checked all the examples and I reviewed all the issues. I could not find any solution.

**Comment by tsv2013** [comment]
Have you loaded select2 library scripts on your page?

Custom widgets + angular are demplstrated in this repo - https://github.com/surveyjs/surveyjs_angular_cli/blob/master/src/app/survey.component.ts

Your code looks ok for me. If you provide us with a live sample issustrating the issue, we could check it.

**Comment by xCozmox** [comment]
@tsv2013 what do you mean exactly with this: 
> Have you loaded select2 library scripts on your page?

I tried this now:
```

import * as widgets from 'surveyjs-widgets';

widgets.select2(Survey);

```

My editor provides me the following "information" it is not an error.: 
Could not find a declaration file for module surveyjs-widgets.

My package.json contains all three:
`        "survey-angular": "^1.0.74",
        "surveyjs-editor": "^1.0.74",
        "surveyjs-widgets": "^1.0.75",`

I'm lost.....

**Comment by andrewtelnov** [comment]
@xCozmox Do you have select2 script on your page? It is a third-party library for our SurveyJS.

Thank you,
Andrew

**Comment by xCozmox** [comment]
@andrewtelnov what do you mean? Yes I did a `npm install select2` Usually there is nothing more to do, or?

I have the following code: 

```

import * as Survey from 'survey-angular';
import * as widgets from 'surveyjs-widgets';

widgets.select2(Survey);

....
....

    loadAll() {
        this.aService
            .query({
                page: 0,
                size: 60
            })
            .subscribe(
                (res: HttpResponse<ICaaQuestion[]>) => this.pushAllQuestions(res.body, res.headers),
                (res: HttpErrorResponse) => this.onError(res.message),
                () => this.createSurvey()
            );
    }

ngOnInit() {   
this.loadAll();  
 this.survey = new Survey.Model();
        this.surveyCountry = this.survey.addNewPage('Country');
        this.surveyCountry1 = this.survey.addNewPage('Country1');
        this.survey.onComplete.add(result => {
            this.surveyCompleted(result.data); 
        });

protected createSurvey() {
        this.surveyCountry.addNewQuestion('dropdown', 'country');
        const countryQuestion = this.survey.getQuestionByName('country');
        countryQuestion.type = 'dropdown';
        countryQuestion.renderAs = 'select2';
        countryQuestion.isRequired = true;
        countryQuestion.title = 'Please select the country:';
        countryQuestion.choices = ['Andorra', 'Germany', 'Italy'];

// afterwards I add the question for Country1

// Survey.Survey.cssType = 'bootstrap';
// Survey.StylesManager.applyTheme('bootstrap');
Survey.SurveyNG.render('surveyElement', { model: this.survey });



```

and the html:
```
``
<div class="row justify-content-center">
    <div class="col-8">
        <form name="editForm" role="form" novalidate (ngSubmit)="save()" #editForm="ngForm">
            <h2 id="jhi-caa-entry-heading">Create or edit a country</h2>
            <div>
                <jhi-alert-error></jhi-alert-error>
                <div id="surveyContainer" class="survey-container contentcontainer codecontainer"><div id="surveyElement"></div></div>
                <jhi-alert></jhi-alert>
                <br/>
        </form>
    </div>
</div>

``
```

**Comment by tsv2013** [comment]
@xCozmox For angular-cli produced applications is not enough to install npm package. In some cases (including the select2 one) you need to add corresponding script in the `.angular-cli.json` file, as you can see in our example - https://github.com/surveyjs/surveyjs_angular_cli/blob/master/.angular-cli.json#L43

**Comment by xCozmox** [comment]
@tsv2013 thanks a lot for your help. 
Unfortunately a jhipster.tech application does not have an .angular-cli.json file. I tried enormous different options. But within the Jhipster application I'm not able to get the select2 option working. If I clone the angluar-cli project it works fine.  I got select2 together with jquery working within jhipster project. But not with surveyjs. I think it is just that I'm missing the js/css links or something like this. The example from the angular-cli project is integraded in my jhipster app. I just looks like missing css. Please find below a print-screen. 
I appreciate any help or hints. I spent hours on it. Thanks a lot!!!
![Screenshot_2019-03-25 SurveyResults](https://user-images.githubusercontent.com/12721026/54925212-21664e00-4f0e-11e9-8ffd-e7f52c638a68.png)

**Comment by tsv2013** [comment]
Unfortunately I'm absolutely nothing know about jhipster.

I've tried to google your case - probably you'll find this helpful - https://github.com/jhipster/generator-jhipster/issues/7941 and https://stackoverflow.com/questions/38473207/how-can-i-using-select2-with-webpack

The very first is the "bruteforce" idea - can you simply add select2 css/js directly on your HTML page?
The next one - can you share your repo for us to take a look what we can do?

**Comment by xCozmox** [comment]
@tsv2013 these 2 solutions don't work for me.
I created an empty jhipster application showing exactly the same issue I have. 
https://github.com/xCozmox/testApp
You can clone it. I didn't touch anything. 
I did:
created an empty entity
created a shared component (webapp/app/shared/survey -  survey and survey-editor component)
npm install select2
npm install survey-angular
npm install surveyjs-editor
npm install @types/jquery

please see the files test-survey.component.html and test-survey.component.ts

It would be great if you could have a look.
mvn and npm start will show you exaclty the same behaviour like I have currently in my application.


Thanks for your help

**Comment by tsv2013** [comment]
@xCozmox I've cloned your repo but haven't found the "files test-survey.component.html and test-survey.component.ts"

Start page doesn't help either:
![image](https://user-images.githubusercontent.com/3013415/55008238-c3ab3200-4ff1-11e9-8c5c-07699813dea8.png)

**Comment by xCozmox** [comment]
@tsv2013 Sorry. I commited only the inital stage of the application. I added the left over. You can pull the changes. 
Important files:
src/main/webapp/entities/survey.component.html
src/main/webapp/entities/survey.component.ts

src/main/webapp/shared/survey/surveyjs.component.ts
src/main/webapp/shared/survey/surveyjs.editor.component.ts

Thanks a lot for your effort.

**Comment by xCozmox** [comment]
@tsv2013 I did an additional commit. I changed the vendor.ts and the vendor.scss in the project to import surveyjs and surveyjs-widgets. This had a positive impact on the the look and feel of surveyjs within jhipster. Now the really only thing which doesn't seem to work is the select2. We are getting closer.

**Comment by tsv2013** [comment]
What page should I navigate to see survey without select2?

**Comment by xCozmox** [comment]
@tsv2013 
Please take my latest push into account. 

just go on Entities--> Survey:
![image](https://user-images.githubusercontent.com/12721026/55062507-65746280-5076-11e9-8fd4-b95a881cd8af.png)

afterwards you see it in question 4 of the survey:
![image](https://user-images.githubusercontent.com/12721026/55065317-e5e99200-507b-11e9-8703-1e8272f8e531.png)

**Comment by tsv2013** [comment]
This should solve the issue - link $+select2 and load select2 CSS

[patch.zip](https://github.com/surveyjs/widgets/files/3016825/patch.zip)

**Comment by tsv2013** [comment]
See
https://stackoverflow.com/questions/38473207/how-can-i-using-select2-with-webpack
https://stackoverflow.com/questions/42069364/how-to-add-custom-js-and-css-files-in-jhipster-4
for more details

**Comment by xCozmox** [comment]
@tsv2013 this worked in the testing application I commited! Thank you so much @tsv2013 
It did not work in the application which I developed.
In this application I build up the survey not via json but via typescript code. And this somehow does not work. 

If I use the following 
`      jQuery(document).ready(() => {
            $('select').select2({ theme: 'bootstrap' });`
I get a select2 box also in my little application. 
The downside of this is that after submitting the form select2 did not select an item out of the dropdownlist. 

I will stop with this for the moment as there seem to be to many dependencies and I'm losing to much time on this. I will develop a normal form instead of using surveyjs. 

Later on I will check again if I can use the components and use them as is via json instead of my code.

Anyway thanks a lot for your help @tsv2013
